lsbi/network.py

```diff
@@ -0,0 +1,202 @@
+        ).squeeze()
+
+    def predict(self, x, alpha=2.0):
+        """Predict the log Bayes Factor.
+
+        log K = lnP(Class 1) - lnP(Class 0)
+        """
+        x = torch.tensor(x, dtype=torch.float32)
+        pred = self.forward(x)
+        pred = self.lpop(pred, alpha=alpha)
+        return pred.detach().numpy()
+
+
+def train(
+    model,
+    data,
+    labels,
+    num_epochs=10,
+    batch_size=128,
+    decay_rate=0.95,
+    lr=0.001,
+):
+    """Train the binary classifier."""
+    # if torch.cuda.is_available():
+    #     device = torch.device("cuda:0")
+    # else:
+    #     device = torch.device("cpu")
+
+    # for now just restrict to cpu
+    device = torch.device("cpu")
+
+    # device = torch.device("mps")
+    print("Using device: ", device)
+
+    # Convert labels to torch tensor
+    data = torch.tensor(data, dtype=torch.float32)
+    labels = torch.tensor(labels, dtype=torch.float32)
+    labels = labels.unsqueeze(1)
+
+    batch_size = batch_size
+    labels = labels.to(device)
+    dataset = torch.utils.data.TensorDataset(data, labels)
+    dataloader = torch.utils.data.DataLoader(
+        dataset, batch_size=batch_size, shuffle=True
+    )
+
+    # Define the loss function and optimizer
+    criterion = model.loss
+
+    optimizer = optim.Adam(model.parameters(), lr=lr)
+
+    # Create the scheduler and pass in the optimizer and decay rate
+    scheduler = ExponentialLR(optimizer, gamma=decay_rate)
+
+    # Train the binary classifier
+    num_epochs = num_epochs
+
+    # Create a DataLoader for batch training
+    model = model.to(torch.float32)
+    # data = data.to(device)
+    model = model.to(device)
+
+    # Turn off the rollback for now as it's likely overkill
+    # best_loss = float("inf")
+    # best_weights = None
+    # patience = 5
+    # counter = 0
+    for epoch in range(num_epochs):
+        epoch_loss = []
+        for i, (inputs, targets) in enumerate(dataloader):
+            # Clear gradients
+            optimizer.zero_grad()
+            inputs = inputs.to(device)
+            # Forward pass
+            loss = criterion(inputs, targets)
+            epoch_loss.append(loss.item())
+            # Backward pass and optimize
+            loss.backward()
+            optimizer.step()
+
+        # Print loss for every epoch
+        scheduler.step()
+        mean_loss = torch.mean(torch.tensor(epoch_loss)).item()
+        print(f"Epoch {epoch+1}/{num_epochs}, Loss: {mean_loss}")
+
+        # if mean_loss < best_loss:
+        #     best_loss = mean_loss
+        #     best_weights = model.state_dict()
+        #     counter = 0
+        # else:
+        #     counter += 1
+        #     if counter >= patience:
+        #         print(
+        #             f"No improvement for {patience} epochs. \
+        #                 Rolling back to best epoch."
+        #         )
+        #         model.load_state_dict(best_weights)
+        #         break
```

Comment:
Do we need this?

Reply:
no, just uncommented to pass coverage but also probably not needed and can be revisited in the future if we do refactor train